compiler/src/iree/compiler/DispatchCreation/FuseHorizontalContractions.cpp

```diff
@@ -283,336 +220,247 @@ static std::optional<HorizontalFusionGroup> getHorizontalFusionGroupMembers(
+                            rhsIndex + 1, linalgOp.getNumDpsInputs() - 1),
+                        [](BlockArgument arg) -> Value { return arg; }));
+
+    appendRange(
+        replacements,
+        llvm::map_range(fusedBody->getArguments().slice(
+                            outsIndex, linalgOp.getNumDpsInits()),
+                        [](BlockArgument arg) -> Value { return arg; }));
+
+    rewriter.mergeBlocks(body, fusedBody, replacements);
+    rhsIndex += linalgOp.getNumDpsInputs() - 1;
+    outsIndex += linalgOp.getNumDpsInits();
+
+    auto yieldOp = cast<linalg::YieldOp>(fusedBody->getTerminator());
+    yieldVals.append(yieldOp->operand_begin(), yieldOp->operand_end());
+    rewriter.eraseOp(yieldOp);
   }
-  Value concatResult = concatResultOp->getResult(0);
-  MutableArrayRef<OpFoldResult> extractSizes(concatResultShape[0]);
-  extractSizes[0] = rewriter.getIndexAttr(1);
-  auto concatResultType = cast<RankedTensorType>(concatResult.getType());
-
-  SmallVector<OpFoldResult> extractOffsets(extractSizes.size(),
-                                           rewriter.getIndexAttr(0));
-  SmallVector<OpFoldResult> extractStrides(extractSizes.size(),
-                                           rewriter.getIndexAttr(1));
-
-  auto concatResultTypeShape =
-      llvm::map_to_vector(concatResultType.getShape(),
-                          [](size_t s) { return static_cast<int64_t>(s); });
-  auto resultOutType =
-      RankedTensorType::get(ArrayRef(concatResultTypeShape).drop_front(),
-                            concatResultType.getElementType());
-
-  SmallVector<Value> replacements;
-  for (auto i : llvm::seq<size_t>(0, rhsValues.size())) {
-    extractOffsets[0] = rewriter.getIndexAttr(static_cast<int64_t>(i));
-    replacements.push_back(rewriter.create<tensor::ExtractSliceOp>(
-        loc, resultOutType, concatResult, extractOffsets, extractSizes,
-        extractStrides));
+  OpBuilder::InsertionGuard g(rewriter);
+  rewriter.setInsertionPointToEnd(fusedBody);
+  rewriter.create<linalg::YieldOp>(loc, yieldVals);
+
+  auto resultsIndex = 0;
+  for (auto linalgOp : linalgOps) {
+    rewriter.replaceOp(linalgOp, fusedOp->getResults().slice(
+                                     resultsIndex, linalgOp->getNumResults()));
+    resultsIndex += linalgOp->getNumResults();
   }
 
-  for (auto [index, op, replacement] :
-       llvm::enumerate(fusionGroup.contractionOps, replacements)) {
-    Operation *replacedOp = op;
-    if (fusionGroup.truncateOps) {
-      replacedOp = fusionGroup.truncateOps.value()[index];
-    }
-    rewriter.replaceOp(replacedOp, replacement);
+  return fusedOp;
+}
+
+static void fuseGroup(RewriterBase &rewriter,
+                      MutableArrayRef<Operation *> fusionGroup,
+                      DominanceInfo &dominanceInfo) {
+  if (!llvm::all_of(fusionGroup, [](Operation *op) {
+        return isa_and_nonnull<linalg::LinalgOp>(op);
+      })) {
+    return;
   }
```

Comment:
I recently learned that we can write it this way, which looks better to me.

```
if (!llvm::all_of(fusionGroup, llvm::IsaPred<linalg::LinalgOp>)) {
  return;
}
```

Reply:
I think `llvm::IsaPred`  doesnt check for the operation not being null (i.e. null value returns true). I want `isa_and_nonnull` (though the op cant be null but just being careful).

Reply:
If it can not be a null, it triggers the assertion in `isa`. So I think it is still cleaner with `IsaPred`. Instead of handling the issue silently, the debug build captures the failure for you.

Reply:
I dont want it to assert though. I want it to return false if it is null.